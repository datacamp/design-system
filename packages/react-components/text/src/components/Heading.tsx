import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { css, SerializedStyles } from '@emotion/core';
import React from 'react';

import baseStyle from '../baseStyle';
import computeDataAttributes from '../computeDataAttributes';
import ssrSafeFirstChildSelector from '../ssrSafeFirstChildSelector';

export type Size = 300 | 400 | 500 | 600 | 700 | 800;
export type HeadingElement = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';

interface HeadingProps {
  className?: string;
  children: string;
  size: Size;
  as: HeadingElement;
  dataAttributes?: { [key: string]: string };
  multiLine?: boolean;
}

const baseHeaderStyle = css({
  fontWeight: tokens.size.fontWeight.bold.value,
  lineHeight: tokens.size.lineHeight.heading.value,
  margin: 0,
  [ssrSafeFirstChildSelector]: {
    marginTop: 0,
  },
});

const singleLineStyle = css({
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
});

const getMarginTop = (size: Size): number => {
  switch (true) {
    case size === 800:
      return tokens.size.spatial.size[24].value;
    default:
      return tokens.size.spatial.size[16].value;
  }
};

const getStyle = (size: Size, multiLine: boolean): SerializedStyles => {
  return css(baseStyle, baseHeaderStyle, multiLine ? {} : singleLineStyle, {
    fontSize: tokens.size.font[size].value,
    marginTop: getMarginTop(size),
  });
};

const Heading: React.FC<HeadingProps> = ({
  children,
  as: Element,
  className,
  size,
  dataAttributes,
  multiLine = false,
}) => {
  const parsedDataAttributes = computeDataAttributes(dataAttributes);
  const style = getStyle(size, multiLine);

  return (
    <Element css={style} className={className} {...parsedDataAttributes}>
      {children}
    </Element>
  );
};

export default Heading;
