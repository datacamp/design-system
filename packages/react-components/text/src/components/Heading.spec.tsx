import 'jest-dom/extend-expect';

import axeRender from '@datacamp/waffles-axe-render';
import React from 'react';

import Heading, { Size } from './Heading';

const testText = 'Some example text.';
const testClassName = 'example-class';
const testDataAttributes = { cy: 'othertest', example: 'testdata' };

const headerSizes: Size[] = [300, 400, 500, 600, 700, 800];

describe('<Heading />', () => {
  describe('renders at each size', () => {
    headerSizes.forEach(size =>
      [true, false].forEach(multiLine => {
        it(`renders at size ${size}, multiLine=${multiLine}`, async () => {
          const { container } = await axeRender(
            <Heading
              as="h2"
              className={testClassName}
              dataAttributes={testDataAttributes}
              multiLine={multiLine}
              size={size}
            >
              {testText}
            </Heading>
          );
          expect(container.firstChild).toMatchSnapshot();
        });
      })
    );
  });

  it('sets the text in the heading', async () => {
    const { getByText } = await axeRender(
      <Heading as="h2" size={400}>
        {testText}
      </Heading>
    );
    expect(getByText(testText)).toBeDefined();
  });

  it('sets the className', async () => {
    const { getByText } = await axeRender(
      <Heading as="h2" className={testClassName} size={300}>
        {testText}
      </Heading>
    );
    expect(getByText(testText)).toHaveClass(testClassName);
  });

  it('uses the correct element', async () => {
    const { container } = await axeRender(
      <Heading as="h5" dataAttributes={testDataAttributes} size={300}>
        {testText}
      </Heading>
    );
    expect(container.querySelector('h5')).not.toBeNull();
  });

  it('sets the data attributes', async () => {
    const { getByText } = await axeRender(
      <Heading as="h2" dataAttributes={testDataAttributes} size={300}>
        {testText}
      </Heading>
    );
    expect(getByText(testText)).toHaveAttribute(
      'data-cy',
      testDataAttributes.cy
    );
    expect(getByText(testText)).toHaveAttribute(
      'data-example',
      testDataAttributes.example
    );
  });
});
