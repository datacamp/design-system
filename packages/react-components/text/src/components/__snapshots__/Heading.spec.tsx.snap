// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> renders at each size renders at size 300, multiLine=false 1`] = `
.emotion-0 {
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-weight: 700;
  line-height: 1.25;
  margin: 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: 16px;
  margin-top: 16px;
}

.emotion-0:first-child:not(style),
.emotion-0 style:first-child + * {
  margin-top: 0;
}

<h2
  class="example-class emotion-0"
  data-cy="othertest"
  data-example="testdata"
>
  Some example text.
</h2>
`;

exports[`<Heading /> renders at each size renders at size 300, multiLine=true 1`] = `
.emotion-0 {
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-weight: 700;
  line-height: 1.25;
  margin: 0;
  overflow: auto;
  white-space: normal;
  font-size: 16px;
  margin-top: 16px;
}

.emotion-0:first-child:not(style),
.emotion-0 style:first-child + * {
  margin-top: 0;
}

<h2
  class="example-class emotion-0"
  data-cy="othertest"
  data-example="testdata"
>
  Some example text.
</h2>
`;

exports[`<Heading /> renders at each size renders at size 400, multiLine=false 1`] = `
.emotion-0 {
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-weight: 700;
  line-height: 1.25;
  margin: 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: 18px;
  margin-top: 16px;
}

.emotion-0:first-child:not(style),
.emotion-0 style:first-child + * {
  margin-top: 0;
}

<h2
  class="example-class emotion-0"
  data-cy="othertest"
  data-example="testdata"
>
  Some example text.
</h2>
`;

exports[`<Heading /> renders at each size renders at size 400, multiLine=true 1`] = `
.emotion-0 {
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-weight: 700;
  line-height: 1.25;
  margin: 0;
  overflow: auto;
  white-space: normal;
  font-size: 18px;
  margin-top: 16px;
}

.emotion-0:first-child:not(style),
.emotion-0 style:first-child + * {
  margin-top: 0;
}

<h2
  class="example-class emotion-0"
  data-cy="othertest"
  data-example="testdata"
>
  Some example text.
</h2>
`;

exports[`<Heading /> renders at each size renders at size 500, multiLine=false 1`] = `
.emotion-0 {
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-weight: 700;
  line-height: 1.25;
  margin: 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: 20px;
  margin-top: 16px;
}

.emotion-0:first-child:not(style),
.emotion-0 style:first-child + * {
  margin-top: 0;
}

<h2
  class="example-class emotion-0"
  data-cy="othertest"
  data-example="testdata"
>
  Some example text.
</h2>
`;

exports[`<Heading /> renders at each size renders at size 500, multiLine=true 1`] = `
.emotion-0 {
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-weight: 700;
  line-height: 1.25;
  margin: 0;
  overflow: auto;
  white-space: normal;
  font-size: 20px;
  margin-top: 16px;
}

.emotion-0:first-child:not(style),
.emotion-0 style:first-child + * {
  margin-top: 0;
}

<h2
  class="example-class emotion-0"
  data-cy="othertest"
  data-example="testdata"
>
  Some example text.
</h2>
`;

exports[`<Heading /> renders at each size renders at size 600, multiLine=false 1`] = `
.emotion-0 {
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-weight: 700;
  line-height: 1.25;
  margin: 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: 24px;
  margin-top: 16px;
}

.emotion-0:first-child:not(style),
.emotion-0 style:first-child + * {
  margin-top: 0;
}

<h2
  class="example-class emotion-0"
  data-cy="othertest"
  data-example="testdata"
>
  Some example text.
</h2>
`;

exports[`<Heading /> renders at each size renders at size 600, multiLine=true 1`] = `
.emotion-0 {
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-weight: 700;
  line-height: 1.25;
  margin: 0;
  overflow: auto;
  white-space: normal;
  font-size: 24px;
  margin-top: 16px;
}

.emotion-0:first-child:not(style),
.emotion-0 style:first-child + * {
  margin-top: 0;
}

<h2
  class="example-class emotion-0"
  data-cy="othertest"
  data-example="testdata"
>
  Some example text.
</h2>
`;

exports[`<Heading /> renders at each size renders at size 700, multiLine=false 1`] = `
.emotion-0 {
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-weight: 700;
  line-height: 1.25;
  margin: 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: 32px;
  margin-top: 16px;
}

.emotion-0:first-child:not(style),
.emotion-0 style:first-child + * {
  margin-top: 0;
}

<h2
  class="example-class emotion-0"
  data-cy="othertest"
  data-example="testdata"
>
  Some example text.
</h2>
`;

exports[`<Heading /> renders at each size renders at size 700, multiLine=true 1`] = `
.emotion-0 {
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-weight: 700;
  line-height: 1.25;
  margin: 0;
  overflow: auto;
  white-space: normal;
  font-size: 32px;
  margin-top: 16px;
}

.emotion-0:first-child:not(style),
.emotion-0 style:first-child + * {
  margin-top: 0;
}

<h2
  class="example-class emotion-0"
  data-cy="othertest"
  data-example="testdata"
>
  Some example text.
</h2>
`;

exports[`<Heading /> renders at each size renders at size 800, multiLine=false 1`] = `
.emotion-0 {
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-weight: 700;
  line-height: 1.25;
  margin: 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: 40px;
  margin-top: 24px;
}

.emotion-0:first-child:not(style),
.emotion-0 style:first-child + * {
  margin-top: 0;
}

<h2
  class="example-class emotion-0"
  data-cy="othertest"
  data-example="testdata"
>
  Some example text.
</h2>
`;

exports[`<Heading /> renders at each size renders at size 800, multiLine=true 1`] = `
.emotion-0 {
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-weight: 700;
  line-height: 1.25;
  margin: 0;
  overflow: auto;
  white-space: normal;
  font-size: 40px;
  margin-top: 24px;
}

.emotion-0:first-child:not(style),
.emotion-0 style:first-child + * {
  margin-top: 0;
}

<h2
  class="example-class emotion-0"
  data-cy="othertest"
  data-example="testdata"
>
  Some example text.
</h2>
`;
